packages/contracts-bedrock/contracts/dispute/AttestationDisputeGame.sol

```diff
@@ -0,0 +1,287 @@
+     *      - When the game resolves, the bond attached to the root claim
+     *      should be distributed among the signers who participated in
+     *      challenging the invalid claim.
+     * @param signature An EIP-712 signature committing to the `rootClaim`
+     *        and `l2BlockNumber` (within the `extraData`) from a key that
+     *        exists within the `attestorSet`.
+     */
+    function challenge(bytes calldata signature) external {
+        // Attempt to recover the signature provided. Solady's ECDSA library
+        // will revert if the signer cannot be recovered from the given signature.
+        address recovered = ECDSA.recoverCalldata(Hash.unwrap(getTypedDataHash()), signature);
+
+        // Check that the recovered address is part of the `attestorSet`.
+        // if (!systemConfig.attestorSet(recovered)) {
+        if (false) {
+            revert InvalidSignature();
+        }
+
+        // If the `recovered` address has already issued a positive
+        // attestation for the `rootClaim`, revert.
+        if (challenges[recovered]) {
+            revert AlreadyChallenged();
+        }
+
+        // Mark that the authorized signer has issued a positive attestation for the `rootClaim`.
+        challenges[recovered] = true;
+
+        // Increment the number of positive attestations that have been issued for the `rootClaim`.
+        attestationSubmitters.push(msg.sender);
+
+        // If the provided signature breaches the signature threshold, resolve the game.
+        if (attestationSubmitters.length == frozenSignatureThreshold) {
+            resolve();
+        }
+    }
+
+    /**
+     * @notice Returns an Ethereum Signed Typed Data hash, as defined in EIP-712, for the
+     *         `Dispute` struct. This hash is signed by members of the `attestorSet` to
+     *         issue a positive attestation for the `rootClaim`.
+     * @return _typedDataHash The EIP-712 hash of the `Dispute` struct.
+     */
+    function getTypedDataHash() public view returns (Hash _typedDataHash) {
+        // Copy the `DISPUTE_TYPE_HASH` onto the stack.
+        Hash disputeTypeHash = DISPUTE_TYPE_HASH;
+        // Grab the root claim of the `AttestationDisputeGame`.
+        Claim _rootClaim = rootClaim();
+        // Grab the L2 block number that the `rootClaim` commits to.
+        uint256 _l2BlockNumber = l2BlockNumber();
+
+        // Hash the `Dispute` struct.
+        Hash disputeStructHash;
+        assembly {
+            // Grab the location of some free memory.
+            let ptr := mload(0x40)
+
+            // Store the `DISPUTE_TYPE_HASH`
+            mstore(ptr, disputeTypeHash)
+            // Store the `rootClaim` of the `AttestationDisputeGame`.
+            mstore(add(ptr, 0x20), _rootClaim)
+            // Store the L2 block number that the `rootClaim` commits to.
+            mstore(add(ptr, 0x40), _l2BlockNumber)
+
+            disputeStructHash := keccak256(ptr, 0x60)
+        }
+
```

Comment:
Do we need to update the free memory pointer here? There is used memory past where the free memory pointer is

Reply:
We should not - we store 96 bytes at the current free memory pointer and hash them, so there will be no references to the data within this memory. It should be completely fine that solc overwrites this memory on the next allocation.

If it'd make you feel better about this, we definitely can though!

Reply:
I suppose its safe as long as other safe code is written, there exists code that this would create a bug for.

Reply:
Got it- not worth keeping track of the possible footgun for 96 bytes of memory expansion, will update the pointer.